Add support for background terminal command execution

hassoncs opened the pull request:
Introduce the `runInBackground` parameter for the `execute_command` tool. This allows commands to be executed without requiring user interaction, which is beneficial for long-running processes such as development servers or build tasks.

The changes include:
- Adding the `runInBackground` parameter to the tool definition and prompt.
- Modifying the `executeCommand` function to handle the `runInBackground` flag, ensuring commands proceed without user approval when set to true.
- Implementing comprehensive unit and integration tests to verify the new functionality.

Discussions:
https://github.com/Kilo-Org/kilocode/discussions/2917
https://github.com/Kilo-Org/kilocode/discussions/2196

I opened the same thing on the Roo side since we haven't sent them anything in a while:
https://github.com/RooCodeInc/Roo-Code/pull/8607

hassoncs commented:
I updated this after discussing this with Hannes on the Roo side. I added more code in order to help the agent manage the running terminal processes. I couldn't think of a better way to do this other than adding a new tool, `terminal_ctrl`.

I thought about maybe adding new parameters to the execute command tool to allow it to send the signal, but that didn't seem very clean either. So, this solution seems to work, but I don't really love the fact that we had to add a tool :-/.

At first, I tried capturing the PIDs of the processes, but it's difficult to get the PID of the built-in VS Code terminal commands. You end up getting the shell process ID, which cannot be killed using the regular `kill` command, which is what the agent tries to do if you don't give it a new tool. So by giving it a tool, we can kill the process similar to the kill button that exists in the UI today. Essentially, this PR gives the agent the ability to click its own buttons in order to continue processes in the background and to kill things later. 

**Test plan**:
I confirmed that the agent can run multiple long-running terminal commands, kill them arbitrarily, and still have visibility into running terminals when dispatching new tasks in orchestrator mode.

hassoncs committed:
feat(terminal): add background command execution and terminal control

- Add `run_in_background` parameter to `execute_command` tool for automatic background execution
- Introduce new `terminal_ctrl` tool for killing processes in terminals
- Update environment details to include terminal ID information for process management
- Refactor tests to focus on kill command generation logic

BREAKING CHANGE: New tool parameters and tools added, may require updates to tool usage patterns

Add change markers

Rename to terminalKill tool

Simplify and revert runInBackground

feat(terminal): refactor listener management for background execution

Adjust listener cleanup logic in terminal processes to support background command execution. Move listener removal from continue() to complete() and abort() methods, and add comprehensive tests for background behavior.

feat(terminal): enable background command execution

Commands now run in detached background processes, allowing long-running tasks to execute without blocking the UI. The terminal integration has been refactored to support non-blocking execution while maintaining output monitoring and proper cleanup.

feat(terminal): add kill requested tracking for terminals

- Add `killRequested` property to `RooTerminal` interface and `BaseTerminal` class
- Set flag to true when kill is initiated and reset on completion or error
- Update environment details to display terminal status as "Killing" or "Active

feat(terminal): introduce process termination tool and optimize execution control

Implement terminal_kill tool for managing active processes with required parameters. Enhance executeCommandTool to prioritize explicit background execution preferences. Streamline ExecaTerminalProcess configuration documentation.

fix(terminal): correct block name formatting in terminal kill case

Remove extra apostrophe from block name string in terminal_kill case to fix formatting inconsistency. Add kilocode_change markers to track modifications in terminal-related code.

feat(terminal): add terminal process kill functionality and background execution support

- Add `runInBackground` and `terminalId` fields to command execution status schema
- Implement `killTerminal` method in TerminalRegistry for targeted process termination
- Refactor terminal kill tool to use registry method and simplify logic
- Update webview message handling to pass terminal ID for operations
- Enhance UI components to display background execution status and kill actions
- Add support for aborting specific terminals via terminal ID

BREAKING CHANGE: Terminal operation handling now requires optional terminalId parameter for targeted control

hassoncs committed:
[chromatic]

chrarnoldus commented on src/core/prompts/tools/execute-command.ts:
```diff
@@ -6,15 +6,18 @@ Description: Request to execute a CLI command on the system. Use this when you n
 Parameters:
 - command: (required) The CLI command to execute. This should be valid for the current operating system. Ensure the command is properly formatted and does not contain any harmful instructions.
 - cwd: (optional) The working directory to execute the command in (default: ${args.cwd})
+- run_in_background: (optional) If true, runs the command in the background without user interaction. Useful for long-running commands like dev servers or monitoring (default: false)
```
should we add it to the native tools as well?

hassoncs committed:
feat(terminal): add terminal kill tool UI and kilocode change markers

- Add kilocode_change markers to terminal schema fields and environment details
- Implement terminal_kill tool UI in ChatRow component with proper markers
- Remove isOrganization prop from payment_required_prompt handling

hassoncs committed:
feat(terminal): add background command execution and terminal kill tool

- Add `run_in_background` parameter to execute_command tool for running commands asynchronously
- Implement terminal_kill tool with improved parameter validation for killing terminal processes
- Update tool registry and execution logic to support new functionality

hassoncs committed:
feat(terminal): enhance terminal management with kill tool and background execution

- Add terminal kill functionality with terminalId support in Task and TerminalRegistry
- Improve background execution cleanup and error handling in ExecaTerminalProcess
- Update terminal kill tool to verify terminal existence and process status
- Enhance UI components for background execution indicators and abort controls
- Add comprehensive tests for race conditions and edge cases in terminal processes
- Refactor terminal process lifecycle management for better reliability

BREAKING CHANGE: Terminal kill tool now requires terminalId parameter for targeted process termination

hassoncs committed:
fix(terminal): correct event emission order and cleanup in ExecaTerminalProcess

- Move 'completed' event emission before removing listeners to ensure proper notification
- Remove unused variables and dead code for cleaner implementation
- Enhance test suite with comprehensive async handling and additional test cases for error scenarios, abort behavior, and background execution

Closes background execution refinements for terminal processes.

eshurakov committed:
Merge main into feature branch

eshurakov committed:
Refactor terminal kill tool implementation

eshurakov committed:
Update tools.ts

eshurakov reviewed (approved):
I've merged latest changes from main and tried it out. Worked fine for me :-)

🚀

hassoncs committed:
Merge branch 'main' into exec-cmd-in-background and resolve conflicts

hassoncs committed:
fix: add missing toolProtocol to terminalKillTool callbacks

hassoncs committed:
Merge remote-tracking branch 'origin/main' into exec-cmd-in-background

hassoncs committed:
Revert unintended changes to main
